File: prodcons1.cpp
// Liam Wynn, 1/19/2021, CS515p: Parallel Programming

/*
  In this assignment, I implement a very simple
  producer-consumer model. It will only have one
  producer and one consumer.
*/

#include <iostream>
#include <thread>
#include <mutex>
#include <condition_variable>
#include "./queue.h"

using namespace std;

// The queue's capacity.
const int BUFSIZE = 20;
// We add all values numbers from 1
// to this value into the queue.
const int NUMITEMS = 100;

Queue queue(BUFSIZE);
mutex mtx_queue_state;
condition_variable queue_not_empty, queue_empty;

// Mutex for controlling output. cout is not atomic!
mutex mtx_output;

void producer() {
  mtx_output.lock();
  cout << "Producer starting on core " << sched_getcpu() << endl;
  mtx_output.unlock();

  for(int i = 1; i <= NUMITEMS; i++) {
    unique_lock<mutex> queue_lck(mtx_queue_state);
    while(!queue.isEmpty())
      queue_empty.wait(queue_lck);

    if(!queue.isFull()) {
      queue.add(i);
      cout << "Producer added " << i << " (qsz: " << queue.size() << ")" << endl;
    }

    queue_not_empty.notify_one();
    queue_lck.unlock();
  }

  mtx_output.lock();
  cout << "Producer ending" << endl;
  mtx_output.unlock();
}

void consumer() {
  mtx_output.lock();
  cout << "Consumer starting on core " << sched_getcpu() << endl;
  mtx_output.unlock();

  for(int i = 1; i <= NUMITEMS; i++) {
    unique_lock<mutex> queue_lck(mtx_queue_state);

    while(queue.isEmpty())
      queue_not_empty.wait(queue_lck);

    int item = queue.remove();
    cout << "Consumer rem'd " << item << " (qsz: " << queue.size() << ")" << endl;

    if(queue.isEmpty())
      queue_empty.notify_one();

    queue_lck.unlock();
  }

  mtx_output.lock();
  cout << "Consumer ending" << endl;
  mtx_output.unlock();
}

int main() {
  thread producer_thread(producer);
  thread consumer_thread(consumer);

  producer_thread.join();
  consumer_thread.join();

  cout << "main: all done!" << endl;
  return 0;
}

File: ProdCons1.java
// Liam Wynn, 1/20/2021, CS515p: Assignment 1

/*
  Implements a very simple producer-consumer model
  in Java. This will only include one producer and
  one consumer.
*/

import java.util.LinkedList;
import java.util.Queue;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.concurrent.locks.Condition;

class ProdCons1 {
  public static final int BUFSIZE = 20;
  public static final int NUMITEMS = 100;

  public static final Lock queueLock = new ReentrantLock();
  public static final Condition queueEmpty = queueLock.newCondition();
  public static final Condition queueNotEmpty = queueLock.newCondition();

  public static void main(String[] args) {
    Queue<Integer> queue = new LinkedList<>();

    Producer producerThread = new Producer(queue);
    Consumer consumerThread = new Consumer(queue);

    try {
      producerThread.start();
      consumerThread.start();

      producerThread.join();
      consumerThread.join();
    } catch(Exception e) {
      System.out.println("Error executing threads");
    }
  }
}

class Producer extends java.lang.Thread {
  private Queue<Integer> queue;

  public Producer(Queue<Integer> queue) {
    this.queue = queue;
  }

  public void run() {
    for(int i = 1; i <= ProdCons1.NUMITEMS; i++) {
      ProdCons1.queueLock.lock();

      try {
        while(queue.size() > 0)
          ProdCons1.queueEmpty.await();

        queue.add(i);
        System.out.println("Added " + i);

        ProdCons1.queueNotEmpty.signal();

      } catch(Exception e) { System.out.println(e.getMessage()); }
      finally {
        ProdCons1.queueLock.unlock();
      }
    }
  }
}

class Consumer extends java.lang.Thread {
  private Queue<Integer> queue;

  public Consumer(Queue<Integer> queue) {
    this.queue = queue;
  }

  public void run() {
    for(int i = 1; i <= ProdCons1.NUMITEMS; i++) {
      ProdCons1.queueLock.lock();

      try {
        while(queue.size() == 0)
          ProdCons1.queueNotEmpty.await();

        int p = queue.remove();
        System.out.println("Rem'd " + p);

        if(queue.size() == 0)
          ProdCons1.queueEmpty.signal();

      } catch(Exception e) { System.out.println(e.getMessage()); }
      finally {
        ProdCons1.queueLock.unlock();
      }
    }
  }
}
